lib: add node:metrics module

Qard opened the pull request:
cc @nodejs/diagnostics 

This is an experiment to see if it would make sense to include a metrics capturing interface in Node.js core. It's only supported with the `node:` prefix as it most definitely conflicts with ecosystem modules with this generic name. We could leave the prefix requirement or rename, I'm not too set on either way.

This takes ideas from various ecosystem metrics modules along with some ideas from the Swift metrics provider. The idea being to provide a generic metrics interface which publishes diagnostics_channel events and allowing consumers to listen to those events and either report them directly or aggregate them in some way.

Qard commented:
I'd have to poke a bit at perf_hooks to figure out how that'd work. I haven't really done anything with it yet. 🤔 

I _was_ hoping to eventually have histograms and other things at some point though, so it _might_ make sense.

legendecas commented on doc/api/metrics.md:
```diff
@@ -0,0 +1,1040 @@
+
+const uniqueUsers = uniqueSet('users.unique');
+
+uniqueUsers.add(user);        // Only counted once per unique value
```
How useful it is to have a `uniqueSet` that's based on the object identity? For two distinct requests, a single user object will be two distinct objects.

Qard replied:
Not sure. That's not the best example though. There's certainly scenarios where object identity could be useful for distinctness.

I mainly just wanted to try the WeakMap approach because just stuffing everything into one giant set forever sounds like a leak to me. 😅

Qard replied:
Could maybe try a Set + WeakSet pair with only numeric types going to the regular Set. It'd be at least _less_ of a leak that way. 🤔 

Also could impose some time window on it and delete old values, even though statsd doesn't appear to do that.

jasnell replied:
Perhaps the `uniqueSet` could accept as an option a callback for overriding the uniqueness determination:

```js
const uniqueUser = createUniqueSet('users.unique', {
  equivalent(a, b) {
    return a.id === b.id;
  }
});

uniqueUser.add({ id: 1 });
uniqueUser.add({ id: 2 });
uniqueUser.add({ id: 1 });
```

Qard replied:
I opted for removing unique set entirely and am instead aiming for value producer wrapper types. I've got Timer now producing values to send to either a gauge or a counter, I'll likely leave it up to ecosystem to provide wrappers for tracking uniqueness counts. The _pattern_ to provide such value producers is in place in any case, though will probably get further polish as I continue to iterate.

legendecas commented on doc/api/metrics.md:
```diff
@@ -0,0 +1,1030 @@
+
+  // Process request...
+
+  timer.stop();
```
The requirement to make a counter from a meter is to augment the counters created with instrumentation library information so that a consumer of the data can tell who produces the data.

This PR assumes that all metrics are created in the global namespace, it is a simplification, but I'd not call the other pattern as odd.

Qard replied:
Not sure why that has to go through a _meter_ for that though. I'm doing essentially the same thing with metadata folding, but I don't need an extra parent type for that. I just have metric-level metadata and report-level data. (And also per-timer data.)

legendecas replied:
A metadata object is just an container and there is no definition what data should be in it, and what an attribute like `api.request.duration` in this PR means, and what unit (milliseconds, or seconds) the duration is recorded with.

OpenTelemetry as an open standard, not only defines the API of the metrics, but also defines the semantics of the metadata attributes: https://github.com/open-telemetry/semantic-conventions/blob/main/docs/README.md.

With the convention defined, a structured API is an abstraction of the data model OpenTelemetry provides.

I don't see how this affects performance. All the metrics objects in OpeTelemetry are pre-created before the hot-path that collects data points. They are just normal JavaScript objects holding extra information.

legendecas reviewed (commented):
I think this module should provide a clear distinguishment between producer of the metrics and consumer of the metrics.

Consumers may be in different collection model, like a push model (OpenTelemetry Collector) and a pull model (Prometheus). This PR provides specific output formats but they still need additional work to properly set up the output.

As a maintainer of OpenTelemetry, I believe that this module should focus on the metrics producer API as a first step.

mcollina commented on doc/api/diagnostics_channel.md:
````diff
@@ -132,6 +132,35 @@ if (diagnostics_channel.hasSubscribers('my-channel')) {
 }
 ```
 
+#### `diagnostics_channel.hasChannel(name)`
````
This is probably worth moving to its own PR.

Qard replied:
Probably, yes. Just was a tiny thing I needed here so added that quickly. 🙂

mcollina commented on lib/metrics.js:
```diff
@@ -0,0 +1,848 @@
+      newMetricChannel.publish(this);
+    }
+
+    ObjectFreeze(this);
```
Why freezing?

Qard replied:
No reason, I just tend to freeze things when they don't need to be mutable. Prevents people accidentally messing them up. 🙈

mcollina commented on lib/metrics.js:
```diff
@@ -0,0 +1,848 @@
+    this.value = value;
+    this.meta = meta;
+    this.time = performance.now();
+    ObjectFreeze(this);
```
why freezing?

jasnell replied:
Exposing these as getters to on the prototype would achieve immutability on the defined members. I don't really see much reason for freezing the entire object.

Qard commented:
> > The intent is for delivery model to be left up to the consumer. It can aggregate in-process and do a pull model, or it can just push data out as the data comes in.
> 
> 
> 
> The current PR is also involved with specific exportation formats.

It only really has those export formats as an experiment, and because it was trivial to write as those formats are very simple.

> > This takes ideas from various ecosystem metrics modules along with some ideas from the Swift metrics provider.
> 
> 
> 
> As mentioned in the OP of the PR, I believe we should clearly define the scope of the module, and leave the exportation to userland libraries. This is also how the mentioned Swift metrics provider does: https://github.com/apple/swift-metrics?tab=readme-ov-file#selecting-a-metrics-backend-implementation-applications-only.

Yes, that's the general intent. We may want to remove the current stream formatters. I mostly included those as I'm trying to figure out to what extent the machinery around that should be included. For example the stream factory could be included, and it's exported currently, to provide a simple interface for people to generate output streams, given that most metric formats fit nicely into that metric-per-line model.

Most likely we'll want to drop the format-specific bits, even though they're trivial, because formats can come and go. We may want to skip the stream factory too though, not sure.

I was also thinking a bit about object model versus channel subscription. In the Swift model there is a class which has a particular protocol which must be confirmed to. We could also do something similar rather than going the channel-based route. It's also possible, as James said, that we could try the perf_hooks base and that may have some implications on the modelling too.

In any case, this is just an experiment at this stage and inviting feedback to guide this toward being something good. I think supporting OpenTelemtry too would be nice, though I worry a bit about imposing too much data requirements on the source. The main concern I have with OTel is that its lack of flexibility due to its significant data presence requirements makes it not so usable with data that's not specifically prepared to match its expectations. This also tends to make that data inflexible to serving other uses as it's so tailored to fit what OTel needs and not necessarily what anyone else needs. 🤔

mcollina reviewed (commented):
Huge +1 on the idea, this is needed - the `prom-client` library introduces a bit too much overhead. I'm not sure about this implementation.

I don't really understand how this is supposed to be used.
Usually, in a prometheus setup, the collector gets all the stats in certain intervals from a given HTTP route. How would the stream based setup work in this case?

Qard commented on lib/metrics.js:
```diff
@@ -0,0 +1,848 @@
+   */
+  #prometheusLabels() {
+    const entries = ObjectEntries(this.meta);
+    const pairs = ArrayPrototypeMap(entries, ({ 0: k, 1: v }) => `${k}="${v}"`);
```
Yep, or more likely removing the formatters entirely and leaving that up to userland. There's probably a bunch more in here than is actually needed. Just did a bunch of experimenting and pushed what I had. 😅

Qard commented:
Yeah, the Prometheus model is a bit different from the other formats I've got in there. I think the pull model can work with just aggregating the data on the other end of the channels and having the pull server read from that, the Prometheus stream is probably a bad example for that. 😅

mcollina commented:
I think we need to be generic enough to support both models correctly.

jasnell commented:
> ... I'd have to poke a bit at perf_hooks to figure out how that'd work.

It really doesn't need to integrate too closely with anything else in `perf_hooks`. Just a `perf_hooks.metrics` would enable this to be semver-minor and able to be backported.

Qard commented:
> It really doesn't need to integrate too closely with anything else in `perf_hooks`. Just a `perf_hooks.metrics` would enable this to be semver-minor and able to be backported.

Ah, gotcha. Easy enough to do. I thought you meant actually integrating with it. Which maybe actually _makes sense_ at least in the timers case. A bit unclear if anything there makes sense for non-timer metrics though. 🤔 

I'll probably poke at this some more maybe next weekend. I think I'm leaning in the direction of detaching this from diagnostics_channel, or at least making that just an optional hook layer, and instead going for an object model to see if I can make this better fit pull models and _possibly_ able to fit better with the data requirements of OpenTelemetry. Like I said, I'm a bit concerned about the verbosity and data requirements OTel might impose, but the wide adoption _should_ be considered. (Though from what I've seen OTel adoption is mainly in tracing with metrics still dominated by other technologies.)

jasnell commented on lib/metrics.js:
```diff
@@ -0,0 +1,848 @@
+   * @private
+   */
+  #statsdType(type) {
+    return {
```
```suggestion
    return {
      __proto__: null,
```

jasnell commented on lib/metrics.js:
```diff
@@ -0,0 +1,848 @@
+  }
+
+  /**
+   * Convert the metric report to a Dogstatsd-compatible string.
```
Would be good to include links to the relevant specs for these in the comments for future generations :-)

Qard replied:
Assuming the formatting bits remain the the final design, certainly. 🙂

jasnell commented on lib/metrics.js:
```diff
@@ -0,0 +1,848 @@
+    if (typeof name !== 'string' || !name) {
+      throw new ERR_INVALID_ARG_TYPE('name', ['string'], name);
+    }
+    if (meta !== undefined && typeof meta !== 'object') {
```
Check for `!== null`?

Qard commented on lib/metrics.js:
```diff
@@ -0,0 +1,848 @@
+    if (!metricTypeNames.includes(type)) {
+      throw new ERR_INVALID_ARG_VALUE('type', type, wrongTypeErr);
+    }
+    if (typeof name !== 'string' || !name) {
```
Yep, I think I did that elsewhere in there but forgot to update that spot. In any case, this will all get a bunch more polish as the full scope of the model gets figured out. 🙂

jasnell commented on lib/metrics.js:
```diff
@@ -0,0 +1,848 @@
+  /**
+   * Reference the timer to prevent to loop from exiting.
+   */
+  ref() {
+    this.#timer?.ref();
+  }
+
+  /**
+   * Unreference the timer to allow the loop to exit.
+   */
+  unref() {
+    this.#timer?.unref();
+  }
```
Would prefer to avoid adding new APIs with `ref()`/`unref()`... these should use `process.ref/unref` (https://nodejs.org/docs/latest/api/process.html#processrefmayberefable)

Qard replied:
Ah, I entirely missed that getting added. Definitely makes sense, especially in this case with timers. Never like that timers didn't quite follow the spec because of the need for ref/unref functions. 😅

jasnell commented on lib/metrics.js:
```diff
@@ -0,0 +1,848 @@
+    this.#interval = interval;
+    this.#fn = fn;
+
+    this.schedule();
```
It would be nice to have an option to disable auto scheduling on construction.

Qard replied:
Maybe? Could also just immediately stop it after construction though, if necessary. 🤔

jasnell commented on lib/metrics.js:
```diff
@@ -0,0 +1,848 @@
+
+    this.#timer = undefined;
+    this.#interval = interval;
+    this.#fn = fn;
```
It would be nice to have an option to perform an initial sample of `fn` on construction

Qard replied:
Probably should be the default, actually. Not sure if one would ever _not_ want an initial sample though, maybe it doesn't even need an option? 🤔

jasnell commented on lib/metrics.js:
```diff
@@ -0,0 +1,848 @@
+    this.duration = this.end - this.start;
+    this.reset(this.duration, mixMeta(this.#meta, meta));
+    return this.duration;
+  }
```
I'd generally prefer things like `end`, `duration`, etc are prototype getters similar to `meta`.

jasnell commented on lib/metrics.js:
```diff
@@ -0,0 +1,848 @@
+   */
+  constructor(metric, interval) {
+    super(metric);
+    if (typeof interval !== 'number' || interval <= 0) {
```
the `interval <= 0` case should probably throw `ERR_INVALID_ARG_VALUE`, likely as a `RangeError`

jasnell replied:
Also, this should probably use the regular validator methods?

jasnell commented on lib/metrics.js:
```diff
@@ -0,0 +1,848 @@
+   * @param {number} value The delta to apply to the gauge.
+   * @param {object} [meta] Additional metadata to include with the report.
+   */
+  applyDelta(value, meta) {
```
should `value` be typed checked or are you counting on / allowing for coercion? If the latter, it would be good to add a comment documenting that

jasnell commented on lib/metrics.js:
```diff
@@ -0,0 +1,848 @@
+   * @param {object} [meta] Additional metadata to include with the report.
+   */
+  reset(value = 0, meta) {
+    this.value = value;
```
type checking for value? Or should `value` be coerced to a number?

jasnell commented on lib/metrics.js:
```diff
@@ -0,0 +1,848 @@
+  increment(n = 1, meta) {
+    if (!this.metric.shouldReport) return;
+
+    if (typeof n === 'object') {
```
should these allow for `null`?

Qard replied:
Not sure there's a reason one would ever provide explicit null there. 🤔

Qard commented on lib/metrics.js:
```diff
@@ -0,0 +1,848 @@
+   * @param {number} [n] The amount to decrement the counter by. Defaults to 1.
+   * @param {object} [meta] Additional metadata to include with the report.
+   */
+  decrement(n = 1, meta) {
```
I'll think on what the appropriate action to take is when there are no subscribers. I was aiming for a similar pattern to diagnostics_channel where publishing without subscribers is a no-op. But users may still want to use the contained values without any export of the data so perhaps you're right. 🤔

jasnell commented on doc/api/metrics.md:
````diff
@@ -0,0 +1,1040 @@
+```
+
+```cjs
+const { counter, timer, statsdStream } = require('node:metrics');
````
I'd prefer the constructor methods here to follow a naming pattern like `createCounter`, `createTimer`, etc, both for ergonomics and for consistency (see `createHistogram(...)` in `perf_hooks`.

```js
const counter = createCounter('api.calls', { service: 'web' });
const timer = createTimer('api.request.duration', { service: 'web' });
```

Qard replied:
My original thinking was to keep the names as simple as possible to be not overly verbose and fairly quick to read, but then in writing tests and docs for it all I found a common case was to want to name the variable the same thing as the function. I may be a bit too radically minimalist about the naming. 😅

jasnell commented on doc/api/metrics.md:
````diff
@@ -0,0 +1,1040 @@
+
+requestRate.mark();     // Mark one event
+requestRate.mark(10);   // Mark 10 events
+```
````
It's unclear from the documentation here how a `meter` differs from a `counter` and what the semantics of calling `mark(...)` multiple times would be. That is, what is the value of the meter after this example.

jasnell commented on doc/api/metrics.md:
```diff
@@ -0,0 +1,1040 @@
+added: REPLACEME
+-->
+
+* {number}
```
Should this also allow for `bigint`?

jasnell commented on doc/api/metrics.md:
```diff
@@ -0,0 +1,1040 @@
+added: REPLACEME
+-->
+
+* `n` {number} The amount to increment. **Default:** `1`
```
Likewise, should `bigint` be supported? Won't mention these further... it applies to all of the new APIs here that accept or yield a `number` currently.

jasnell commented on doc/api/metrics.md:
````diff
@@ -0,0 +1,1040 @@
+apiCalls.increment(5);    // Increment by 5
+apiCalls.increment(10, { endpoint: '/api/users' }); // Increment by 10 with metadata
+apiCalls.increment({ endpoint: '/api/orders' });    // Increment by 1 with metadata
+```
````
If it does not do so currently, `increment` should probably throw if the `value` is negative. Likewise with `decrement`.

jasnell commented on doc/api/metrics.md:
```diff
@@ -0,0 +1,1040 @@
+import { stdout } from 'node:process';
+
+const stream = dogstatsdStream();
+stream.pipe(stdout);
```
It doesn't need to be figured out in the initial version here, but it would be good to document examples of filtering these streams or passing them through transforms. Also, how are you imagining sampling to be implemented?

jasnell commented on doc/api/metrics.md:
```diff
@@ -0,0 +1,1040 @@
+* `n` {number} The number of events to mark. **Default:** `1`
+* `meta` {Object} Additional metadata for this report.
+
+Records events and updates the rate calculation.
```
Having a meter of marks per time interval is great but I wonder if it would be worthwhile to have a meter of marks per other unit as well. For instance, something like...

```
const rate = createMeter('marks.per.tick');
rate.mark(10);
rate.tick();  // 10 marks / tick
rate.mark(20);
rate.tick(); // 20 marks / tick
// etc
```

jasnell commented on doc/api/metrics.md:
````diff
@@ -0,0 +1,1040 @@
+errorCount.increment();     // Increment by 1
+errorCount.increment(5);    // Increment by 5
+errorCount.decrement(2);    // Decrement by 2
+```
````
For most of these, it would also be helpful for the metric to provide a counter for the number of discreet observations made. E.g. for `counter`, it would be the number of times `increment`/`decrement` are called, etc. For an example of how this could be used, on `uniqueSet`, it can be used to calculate the percentage of observations that end up being unique or duplicated

legendecas commented on doc/api/metrics.md:
```diff
@@ -0,0 +1,1040 @@
+
+* Returns: {string}
+
+Formats the metric report as a DogStatsD-compatible string with tags.
```
I'm concerned about integrating a vendor specific format into the built-in API, not mentioning this is a commercial vendor.

jasnell replied:
Yeah, personally I'd prefer to have our API publish a generic stream using our own format with a documented example showing how to adapt that to specific formats using transforms.

legendecas replied:
Agreed, I think the scope of this `node:metrics` module should be providing a generic metric producer API, rather than providing a exportation format, as I also requested at https://github.com/nodejs/node/pull/58874#pullrequestreview-2968963753.

But Node.js does not provide a metric backend. I'm not sure how valuable it is to invent a output format without a backend support.

Qard replied:
Just an FYI: while dogstatsd was _created_ by Datadog, it's quite commonly used elsewhere. Many languages have modules supporting it which are in no way associated with Datadog, like Node.js has hotshots.

But yes, the formats are almost certainly all getting deleted before landing, that was just me experimenting with the model to see that it could serve many formats well.

Please don't consider any of what's here now as intended for the final product, this is all experiments to see what works and what doesn't do define what the _real_ final design will be.

legendecas commented on test/parallel/test-metrics-meter.js:
```diff
@@ -0,0 +1,43 @@
+
+const messages = [
+  [1, { base: 'test' }],
+  [124, { base: 'test', meta: 'extra' }],
```
Is it expected that the value of the meta `{ base: 'test', meta: 'extra' }` is `124`, while it was only `mark`-ed with `123`?

Qard replied:
Yes, it's cumulative within the specified window.

legendecas replied:
The meta `{ base: 'test', meta: 'extra' }` is only marked once with value `123` though.

Qard replied:
The value is not unique to the meta, it is unique to the instance. That single data point report just has additional data on it. You can have type-level meta and you can have add report-level meta onto individual reports, but the value is always based on the type instance.

legendecas replied:
> In the case of the meta folding, I'm thinking of instead of having one meta in the final reporting, keeping the instance and report metas separate so the consumers can make their own decisions on what to do with the different metadata sets.

From the current PR status, the aggregation is already done inside the `Meter.mark`. Is this describing the goal of the API, rather than the current PR status?

Qard replied:
Yes, that's ideation on a possible improvement to make the distinction between the two meta stages and to allow taking the data from one and not the other if a particular format doesn't support it.

legendecas commented on doc/api/metrics.md:
```diff
@@ -0,0 +1,1040 @@
+* `meta` {Object} Optional metadata to attach to all reports.
+* Returns: {metrics.Meter}
+
+Creates a meter metric that measures the rate of events over time.
```
What does the "rate" here refer to? I think this meter type needs more explanation as it is not a common type in existing metric backends.

Qard replied:
Ah, I misunderstood, it's actually not part of vanilla statsd apparently, but rather is included as an addition in basically every implementation of statsd. In dogstatsd it's called rate, in Sysdig it's called meter, most other APM vendors have some variation of it.

In the case of Prometheus, it doesn't make sense because the pull model keeps the data in-app, and having that unaggregated would be prohibitively expensive.

legendecas replied:
This sounds like the API design should still be a counter. In the exportation, they can decide if they would like to aggregate in a time interval, or until being pull-ed.

Qard replied:
They are generally marked as separate types right from the initial API as the backend usually needs intent communicated to know which aggregation method to apply. Some just apply all the aggregations though and make every representation available, though that can get very expensive for storage.

legendecas replied:
To clarify my comment at https://github.com/nodejs/node/pull/58874#discussion_r2174902336. I think it is fine to add more semantic attributes to a metric. If a backend needs more semantics, they can even be added as meta.

> Applying an interval and aggregation in the metric collection adds additional hurdles when there are exporters configured with different exportation intervals. They have to handle the aggregation with different size of time windows.

What's really concerning is doing aggregation in the metric collection. This practically prohibits some backend exporters to be implemented efficiently.

vicb commented on doc/api/metrics.md:
```diff
@@ -0,0 +1,1040 @@
+* `meta` {Object} Optional metadata to attach to all reports.
+* Returns: {metrics.TimerFactory}
+
+Creates a timer factory for measuring durations.
```
I guess those have the same limitations as any timers in the runtime i.e. https://developers.cloudflare.com/workers/reference/security-model/#step-1-disallow-timers-and-multi-threading?

Qard replied:
It's based on performance.now(), not Date.now(), which is a monotonic counter since process start, so would be correct.

vicb replied:
Oops sorry, James added a message on our internal chat, I thought I was looking at some cloudflare pr 🤦‍♂️

jasnell commented:
> (fwiw i'd strongly prefer a rename so the prefix isn't required; prefix-only modules are a complexity for tooling)

I mentioned this higher up in the thread, but I think we can avoid using a new top-level module for this entirely by exposing it via `perf_hooks.metrics`. That would make it semver-minor and backportable and side steps some of these other issues.

legendecas commented on doc/api/metrics.md:
```diff
@@ -0,0 +1,1040 @@
+added: REPLACEME
+-->
+
+A metric that only increases or decreases.
```
It is important to add a monotonic counter (increase only) to allow backends like Prometheus to save this metric correctly.

Qard replied:
Yep, going to get to that. I'm working on figuring out a good modelling for having multiple counter types to be able to express that effectively.

Qard commented on doc/api/metrics.md:
```diff
@@ -0,0 +1,1040 @@
+
+const uniqueUsers = uniqueSet('users.unique');
+
+uniqueUsers.add(user);        // Only counted once per unique value
```
Yeah, I just removed it. That can be a follow-up thing, or even just let userland handle that. I'm going for a new modelling with case-specific things like this using an additional modelling layer over raw metric types to produce values that core metrics can use. The first iteration of this idea being Timer, which is now just reduces a start and end to a duration which becomes an integer value which any of the metric types can consume. Similarly, a unique set would probably just be a container type which applies whatever uniqueness logic it wants to then trigger a producer function that routes value changes into whichever metric it's bound to. The concept still might need work, but that should provide both flexibility while also removing responsibility for core to implement every possible thing.

Qard commented:
I made a bunch of changes, but there's still more to do. I cleaned up the object model for the producing side, but there's still much I still need to figure out on the consuming side, particularly with how to wire up pull-based types. I replaced `PeriodicGauge` with a `PullGauge` type which should be possible to use both in a pull mechanism and in a in-process aggregating timed push model.

legendecas commented on doc/api/metrics.md:
```diff
@@ -0,0 +1,635 @@
+
+// Create counter metrics
+const apiCalls = createCounter('api.calls', { service: 'web' });
+const requestDuration = createCounter('api.request.duration.ms', { service: 'web' });
```
Using counters to record durations is not a good practice in real life. It is not a single counter value been incremented. It should be represented by a distribution of values, like Histogram.

The API should distinguish between counters and histograms.

Qard commented on doc/api/metrics.md:
````diff
@@ -0,0 +1,635 @@
+memory.reset(memoryUsage().heapUsed);
+```
+
+### `metrics.createPullGauge(name, fn[, meta])`
````
At the moment it needs some interaction to get the value out, so a pull is required. Not yet sure of a good way to support both push and pull with the same data type, trying to figure that out.

RafaelGSS commented on lib/internal/bootstrap/realm.js:
```diff
@@ -124,6 +124,7 @@ const legacyWrapperList = new SafeSet([
 // beginning with "internal/".
 // Modules that can only be imported via the node: scheme.
 const schemelessBlockList = new SafeSet([
+  'metrics',
```
Since we will expose this API under perf_hooks, this doesn't seem necessary anymore

legendecas commented on doc/api/metrics.md:
```diff
@@ -0,0 +1,1023 @@
+-->
+
+* `name` {string} The metric name. Must not be empty.
+* `options` {Object}
```
The ability to describe monotonicity is important for backends like Prometheus to correctly process the metric data.

Missing the intent of a metric monotonicity would make Prometheus consumers not been able to use functions like `rate`, `increase` etc.

The producer of the metric should be responsible for expressing the intent of the monotonicity.

legendecas commented:
> It describes it at the consumer end, that's the whole point. The producer side just decides the semantics of what the numbers it is recording are about and the consumer decides how to consume that sequence of numbers. It can produce a counter or gauge if it wants, or it can get bucketed into histograms. The producer doesn't need to care about the storage or wire optimization mechanisms you might want to apply to reduce the data, it just cares about saying "here's a data point" to the consumers.

Metric value semantics are essential to be described at the producer end to accurately express the intent. Inferring metric properties like monotonicity and value type at consumer end would result in accuracy loss in the value translations.

If a producer produces both integer and float point numbers for a single metric, and a consumer infers it as an integer metric because the first number is an integer, this creates accuracy loss. If a produce only monotonically increments a counter, there is no way for a consumer to infer its monotonicity and exports it correctly. Inferring the metric data by metric name is not a programmatically optimal metric API.

These are lessons learnt in OpenTelemetry API design and [specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#counter) already. The expressiveness of the metric API is critical to a lossless vendor-neutral exportation.

legendecas commented:
Assuming https://github.com/nodejs/node/pull/61907 lands the OpenTelemetry tracing support, and OpenTelemetry also supports metrics, an integral support with both OTel tracing and metrics would avoid discrepancy and separation.

I'm not advocating including the @opentelemetry/api package as-is now. As @dyladan mentioned in https://github.com/nodejs/node/pull/61907#issuecomment-3946362610, the OpenTelemetry JS API @opentelemetry/api is been experimented with an event-based (or `diagnostic_channels`-based) design as well.

OpenTelemetry metrics [data model](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/data-model.md) defines a similar data pipeline like the API emitting metric events and the SDK aggregates the data events:

![data-model](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/img/model-layers.png?raw=true)

With OpenTelemetry's vendor-neutral design as an existing art, I think we should take advantage of it, and I would agree that we could improve the implementation like adopting `diagnostic_channels`.

legendecas commented:
> It is NOT the intent to just ship specifically what OpenTelemetry has decided metrics should look like, because much of the market does not agree. It is ONLY the intent to allow OpenTelemetry to pull data in the structure it expects as one possible interpretation.

Regardless of this API to follow OpenTelemetry or not, I would like to argue that vendor-neutrality is a hard requirement for it to be included in Node.js. The API should be balanced to be able for adoption of various vendors with accurate lossless data exportation.

OpenTelemetry is not a vendor, but as an open standard has been adopted by vast [majority vendors](https://opentelemetry.io/ecosystem/vendors/).

Qard commented:
@nodejs/diagnostics Anything that can be done to keep this moving? It could use more feedback.

jsumners-nr commented:
Honestly, I just don't have enough experience with metrics systems to provide much useful feedback. I read through most of the new docs, and I'd recommend expanding on the `unit` docs. I have no idea what a `By` unit is, how it is supposed to be used, and does it require the "observable" function to operate correctly? The description of "observable" indicates it is for something else. Also, it's not clear to me why `{count}` is different from the other values. It looks like a template value, whereas the others are clearly static.
